lib/SyTest/Federation/Room.pm

```diff
@@ -0,0 +1,138 @@
+package SyTest::Federation::Room;
+
+use strict;
+use warnings;
+
+sub make_event_refs
+{
+   [ map { [ $_->{event_id}, $_->{hashes} ] } @_ ];
+}
+
+sub create
+{
+   my $class = shift;
+   my %args = @_;
+
+   my $server = $args{server};
+
+   my $creator = $args{creator};
+   my $room_id = $args{room_id} // $server->next_room_id;
+
+   my $self = bless {
+      room_id => $room_id,
+      server  => $server,
+
+      current_state => {},
+      prev_events => [],
+   }, $class;
+
+   $self->create_event(
+      type => "m.room.create",
+
+      content     => { creator => $creator },
+      depth       => 0,
+      sender      => $creator,
+      state_key   => "",
+   );
+
+   $self->create_event(
+      type => "m.room.member",
+
+      content     => { membership => "join" },
+      depth       => 0,
+      sender      => $creator,
+      state_key   => $creator,
+   );
+
+   $self->create_event(
+      type => "m.room.join_rules",
+
+      content     => { join_rule => "public" },
+      depth       => 0,
+      sender      => $creator,
+      state_key   => "",
+   );
+
+   return $self;
+}
+
+sub room_id
+{
+   return $_[0]->{room_id};
+}
+
+sub create_event
+{
+   my $self = shift;
+   my %fields = @_;
+
+   my $server = $self->{server};
+
+   $fields{prev_state} = [] if defined $fields{state_key}; # TODO: give it a better value
+
+   my @auth_events = grep { defined }
+      $self->get_current_state_event( "m.room.create" ),
```

Comment:
()s around the array you're ?implicitly constructing would be nice, if they're allowed